src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,116 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] elementData;
+    private int size;
+
+    public ArrayList() {
+        elementData = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == elementData.length) {
+            increaseCapacity();
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException(indexOutOfBoundsMsg(index));
+        }
+        if (!(size < elementData.length)) {
+            increaseCapacity();
+        }
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list != null) {
+            for (int i = 0; i < list.size(); i++) {
+                add(list.get(i));
+            }
+        }
+    }
 
+    private void increaseCapacity() {
+        int newCapacity = size + (size >> 1);
+        Object[] oldElementData = elementData;
+        elementData = new Object[newCapacity];
+        System.arraycopy(oldElementData, 0, elementData, 0, oldElementData.length);
     }
 
     @Override
+    @SuppressWarnings("unchecked")
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elementData[index] = value;
     }
 
     @Override
+    @SuppressWarnings("unchecked")
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        final T removedElement = (T) elementData[index];
+        int numberOfElementsToMove = size - index - 1;
+        if (numberOfElementsToMove > 0) {
+            System.arraycopy(elementData, index + 1, elementData, index, numberOfElementsToMove);
+        }
```

Comment:
```suggestion
          System.arraycopy(elementData, index + 1, elementData, index, numberOfElementsToMove);
```
it won't do anything in that case anyway